client/executor/src/wasm_runtime.rs

```diff
@@ -221,7 +240,17 @@ impl RuntimeCache {
 		let code_hash = &runtime_code.hash;
 		let heap_pages = runtime_code.heap_pages.unwrap_or(default_heap_pages);
 
-		let mut runtimes = self.runtimes.lock(); // this must be released prior to calling f
+		// this must be released prior to calling f.
+		let mut runtimes = match runtime_code.context {
+			sp_core::traits::CodeContext::Consensus => self.consensus_runtimes.lock(),
+			sp_core::traits::CodeContext::Offchain => self.offchain_runtimes.lock(),
```

Comment:
So offchain could not use consensus instance, since heappage differs I guess it would not be a good idea to try.
I do like that there is a clean separation.